### ISSUE TITLE
[Web] Performance regression in 20260502 version when using intel xe-lpg iGPU

### ISSUE THREAD

**Issue by xmcp** [microsoft/onnxruntime#28531]
### Describe the issue

When using the iGPU of Core Ultra 185H, onnxruntime-web `1.26.0-dev.20260502-9d1492a449` shows 4x worse performance over the previous dev build, `1.26.0-dev.20260416-b7804b056c`. This regression continues to exist in `1.26.0` stable and also the latest dev `1.27.0-dev.20260506-673c3320fc`.

I used Codex to write a minimal reproduction script (see the next section). Here is the result, where the inference time of the `Xenova/bge-base-zh-v1.5` model increases from 529.0ms to 2234.9ms:

<img width="608" height="292" alt="Image" src="https://github.com/user-attachments/assets/19132bba-b606-4ae0-b585-0231cd78fbd0" />

In comparison, I see no significant regression on RTX 4060 Laptop:

<img width="608" height="292" alt="Image" src="https://github.com/user-attachments/assets/16076294-16df-4bc2-989f-f2c2c8ab1784" />

So this regression should specific to intel.

### To reproduce

Save this as HTML, open it, and click Run.

```html
<!doctype html>
<button id="run">Run</button>
<pre id="out"></pre>
<script type="module">
const modelUrl = "https://huggingface.co/Xenova/bge-base-zh-v1.5/resolve/main/onnx/model_fp16.onnx";
let cachedModelUrlPromise;
const cases = [
  ["1.26.0-dev.20260416-b7804b056c"],
  //["ci-20260425-954dbce243", ".ort-ci/954dbce243/dist/"],
  ["1.26.0-dev.20260502-9d1492a449"],
  ["1.26.0"],
  ["1.27.0-dev.20260506-673c3320fc"],
];
const batch = 128, seq = 32, runs = 3, warmup = 1;

async function gpuInfo() {
  const adapter = await navigator.gpu?.requestAdapter();
  if (!adapter) return "GPU: unavailable";
  const info = adapter.info ?? await adapter.requestAdapterInfo?.();
  return `GPU: ${info?.vendor ?? "unknown vendor"} ${info?.architecture ?? ""} ${info?.device ?? ""} ${info?.description ?? ""}`.trim();
}

function feeds(ort) {
  const n = batch * seq;
  const ids = new BigInt64Array(n);
  const zeros = new BigInt64Array(n);
  const mask = new BigInt64Array(n).fill(1n);
  for (let b = 0; b < batch; b++) {
    const row = b * seq;
    ids[row] = 101n;
    ids[row + seq - 1] = 102n;
    for (let s = 1; s < seq - 1; s++) ids[row + s] = BigInt(100 + ((b + s) % 1000));
  }
  const dims = [batch, seq];
  return {
    input_ids: new ort.Tensor("int64", ids, dims),
    attention_mask: new ort.Tensor("int64", mask, dims),
    token_type_ids: new ort.Tensor("int64", zeros, dims),
  };
}

async function getCachedModelUrl() {
  cachedModelUrlPromise ??= fetch(modelUrl)
    .then((response) => {
      if (!response.ok) throw new Error(`failed to fetch model: ${response.status} ${response.statusText}`);
      return response.blob();
    })
    .then((blob) => URL.createObjectURL(blob));
  return cachedModelUrlPromise;
}

async function bench([name, distBase]) {
  const base = distBase ?? `https://cdn.jsdelivr.net/npm/onnxruntime-web@${name}/dist/`;
  const suffix = `?case=${encodeURIComponent(name)}`;
  const ort = await import(`${base}ort.webgpu.min.mjs${suffix}`);
  ort.env.logLevel = "error";
  ort.env.wasm.wasmPaths = {
    wasm: `${base}ort-wasm-simd-threaded.asyncify.wasm`,
    mjs: `${base}ort-wasm-simd-threaded.asyncify.mjs`,
  };
  const session = await ort.InferenceSession.create(await getCachedModelUrl(), {
    executionProviders: [{ name: "webgpu", validationMode: "disabled" }],
    graphOptimizationLevel: "all",
  });
  const input = feeds(ort);
  for (let i = 0; i < warmup; i++) await session.run(input);
  let total = 0;
  for (let i = 0; i < runs; i++) {
    const t0 = performance.now();
    const output = await session.run(input);
    output[session.outputNames[0]].cpuData[0];
    total += performance.now() - t0;
  }
  await session.release?.();
  return `${name}: ${(total / runs).toFixed(1)} ms`;
}

document.querySelector("#run").onclick = async () => {
  const out = document.querySelector("#out");
  out.textContent = `${await gpuInfo()}\n`;
  await getCachedModelUrl();
  out.textContent += "running...\n";
  for (const c of cases) out.textContent += `${await bench(c)}\n`;
  out.textContent += "done.\n";
};
</script>
```

*(20260519: updated the script to cache the .onnx file into blob url, instead of downloading it once per test case)*

### Urgency

N/A

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.26.0-dev.20260502-9d1492a449

### Execution Provider

'webgpu' (WebGPU)

**Comment by xmcp** [comment]
I cannot further bisect the issue because there are no more intermediate builds on npm, and CI builds on GitHub are already expired. I can help to bisect if more intermediate builds are available.

**Comment by hariharans29** [comment]
Do you see the perf regression when batch size == 1 ?

There was a PR to enable split k when batch size > 1: https://github.com/microsoft/onnxruntime/pull/28151 (landed within the regression time window). If the regression goes away with making batch size == 1, that makes that PR a strong suspect. Else, we can move on to other candidates.

Alternatively, if you revert that PR locally and build from source, that would give a definitive answer too.

CC: @guschmue

**Comment by Jiawei-Shao** [comment]
I will look into this issue. Thanks for your report!

**Comment by xmcp** [comment]
Yes the problem goes away with `const batch = 1, seq = 128, runs = 3, warmup = 1;`:

<img width="609" height="265" alt="Image" src="https://github.com/user-attachments/assets/3b9e12c5-a75b-4c57-829f-78391719d9a2" />

If someone can send me a built mjs/wasm with that PR reverted I can help to test.

**Comment by Jiawei-Shao** [comment]
I can also reproduce the regression with the webpage in Chrome on my Meteor Lake iGPU.

I am debugging ORT to understand more about it.

**Comment by Jiawei-Shao** [comment]
In fact I don't think the regression is related to [28151](https://github.com/microsoft/onnxruntime/pull/28151) because now we only support Split-K when [batch size <= 8](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/webgpu/webgpu_utils.cc#L55), while in this case the batch size is 128, and I don't see Split-K is used in the dumped shaders.

**Comment by hariharans29** [comment]
Could it be related to this: https://github.com/microsoft/onnxruntime/pull/28197 ?

**Comment by Jiawei-Shao** [comment]
> Could it be related to this: https://github.com/microsoft/onnxruntime/pull/28197 ?

I don't see any obvious performance changes after reverting https://github.com/microsoft/onnxruntime/pull/28197 with onnxruntime_perf_test.

**Comment by xmcp** [comment]
Hi, I have bisected this regression by building from source, and the culprit is indeed 6f47410e10e968e398fa65daea4866d1d23c9d01 (#28197). Performance drops since that commit:

<img width="769" height="380" alt="Image" src="https://github.com/user-attachments/assets/4ed42493-385d-4159-b8e3-44f3ad93d64f" />

cc @xhcao

-----

> I don't see any obvious performance changes after reverting #28197 with onnxruntime_perf_test.

Does onnxruntime_perf_test use WebGPU? This regression is specific to the web platform (more accurately, specific to Chrome*) so the perf test may not catch it.

[*]: Below is the test result with Firefox 149. You can see there is no performance regression, but it is overall 3x slower than Chrome for all versions. A possible explanation is that Firefox returns empty vendor info, effectively bypassing all intel-specific optimizations.

<img width="937" height="691" alt="Image" src="https://github.com/user-attachments/assets/05db2942-f3e6-46d5-8ec7-d7e83763bbc8" />

**Comment by xhcao** [comment]
Hi, @xmcp, I will investigate the issue.

**Comment by xhcao** [comment]
https://github.com/microsoft/onnxruntime/pull/28760 @xmcp I had drafted a PR, and verified that it could fix your issue. Could you help verify it on your machine again. Thanks.
Please verify it with different batch size and sequence length.

**Comment by xmcp** [comment]
I tried multiple settings and this PR did fix the regression on my Core Ultra 185H. :clap:

| setup              | 1.26.0-dev.20260416-b7804b056c | 1.26.0-dev.20260502-9d1492a449 | main_a8baf5c4a5 |  pr_28760 |
| :----------------- | -----------------------------: | -----------------------------: | --------------: | --------: |
| batch=1, seq=4     |                        25.3 ms |                        17.9 ms |         19.0 ms |   16.6 ms |
| batch=1, seq=13    |                        25.1 ms |                        23.7 ms |         18.0 ms |   18.1 ms |
| batch=1, seq=32    |                        23.9 ms |                        23.2 ms |         23.8 ms |   23.5 ms |
| batch=1, seq=57    |                        43.1 ms |                        31.3 ms |         31.8 ms |   30.5 ms |
| batch=1, seq=128   |                       108.3 ms |                        92.1 ms |         92.3 ms |   53.9 ms |
| batch=4, seq=4     |                        24.6 ms |                        25.6 ms |         17.1 ms |   17.5 ms |
| batch=4, seq=13    |                        31.3 ms |                        42.2 ms |         30.8 ms |   30.4 ms |
| batch=4, seq=32    |                        42.9 ms |                       105.1 ms |         92.4 ms |   43.1 ms |
| batch=4, seq=57    |                       110.8 ms |                       102.1 ms |         98.9 ms |   50.1 ms |
| batch=4, seq=128   |                       187.5 ms |                       186.5 ms |        188.6 ms |   77.4 ms |
| batch=13, seq=4    |                        44.2 ms |                        30.6 ms |         30.1 ms |   30.2 ms |
| batch=13, seq=13   |                        49.3 ms |                        89.4 ms |         89.1 ms |   58.9 ms |
| batch=13, seq=32   |                        65.4 ms |                       161.8 ms |        160.9 ms |   65.0 ms |
| batch=13, seq=57   |                       283.6 ms |                       254.4 ms |        253.5 ms |  113.6 ms |
| batch=13, seq=128  |                       837.0 ms |                       829.5 ms |        838.8 ms |  223.0 ms |
| batch=32, seq=4    |                        59.2 ms |                        91.2 ms |         91.3 ms |   44.3 ms |
| batch=32, seq=13   |                        76.4 ms |                       164.6 ms |        161.7 ms |   67.4 ms |
| batch=32, seq=32   |                       131.4 ms |                       376.1 ms |        375.5 ms |  134.0 ms |
| batch=32, seq=57   |                       929.0 ms |                       870.9 ms |        925.1 ms |  300.4 ms |
| batch=32, seq=128  |                      1994.2 ms |                      2141.3 ms |       2133.3 ms |  526.3 ms |
| batch=57, seq=4    |                        78.5 ms |                       100.2 ms |         99.1 ms |   51.6 ms |
| batch=57, seq=13   |                       122.8 ms |                       251.0 ms |        251.7 ms |  111.3 ms |
| batch=57, seq=32   |                       230.0 ms |                       867.1 ms |        866.8 ms |  229.9 ms |
| batch=57, seq=57   |                      1734.2 ms |                      1658.7 ms |       1640.5 ms |  435.7 ms |
| batch=57, seq=128  |                      3703.4 ms |                      4187.3 ms |       4197.2 ms |  911.1 ms |
| batch=128, seq=4   |                       164.0 ms |                       184.8 ms |        185.8 ms |   76.0 ms |
| batch=128, seq=13  |                       268.4 ms |                       823.5 ms |        826.1 ms |  221.2 ms |
| batch=128, seq=32  |                       507.5 ms |                      2135.4 ms |       2109.4 ms |  515.4 ms |
| batch=128, seq=57  |                      3983.9 ms |                      4235.2 ms |       4231.1 ms |  944.3 ms |
| batch=128, seq=128 |                      8310.8 ms |                      9516.4 ms |       9633.2 ms | 2040.8 ms |

A better visualization below:

<img width="947" height="520" alt="Image" src="https://github.com/user-attachments/assets/d7a69449-4caa-4a44-a62b-8de029738cdf" />

As a result, the performance with this PR is generally on par with the dev.20260416 baseline when batch/seq are small, or significantly faster than the baseline as they increase. So, nice work :)
